Check that `#[may_dangle]` is properly applied

GoldsteinE opened the pull request:
It's only valid when applied to a type or lifetime parameter in `Drop` trait implementation.

Tracking issue: https://github.com/rust-lang/rust/issues/34761
cc https://github.com/rust-lang/rust/issues/34761#issuecomment-1208185551

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10441003742/job/28911603273) [(plain)](https://github.com/rust-lang/rust/commit/f1febc1e0c66578dae6a4ad074f397b679ff5377/checks/28911603273/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.660 Building wheels for collected packages: reuse
#13 2.661   Building wheel for reuse (pyproject.toml): started
#13 2.907   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.908   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.908   Stored in directory: /tmp/pip-ephem-wheel-cache-j4wyic0n/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.911 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.316 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.317 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.845 Collecting virtualenv
#13 3.845 Collecting virtualenv
#13 3.885   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 3.997      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 51.9 MB/s eta 0:00:00
#13 4.053 Collecting filelock<4,>=3.12.2
#13 4.057   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.077 Collecting distlib<1,>=0.3.7
#13 4.081   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.090      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 68.2 MB/s eta 0:00:00
#13 4.119 Collecting platformdirs<5,>=3.9.1
#13 4.123   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.206 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.396 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      200640 kB
DirectMap2M:     8187904 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/334e509912d82323ff0fca016a2e69dbbee559fd/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-334e509912d82323ff0fca016a2e69dbbee559fd-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 30.01s
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_passes/src/check_attr.rs:1377:
     fn check_may_dangle(&self, hir_id: HirId, attr: &Attribute, target: Target) {
         if matches!(target, Target::GenericParam(_))
             && let hir::Node::GenericParam(param) = self.tcx.hir_node(hir_id)
-            && matches!(param.kind, hir::GenericParamKind::Lifetime { .. } | hir::GenericParamKind::Type { .. })
+            && matches!(
+                param.kind,
+                hir::GenericParamKind::Lifetime { .. } | hir::GenericParamKind::Type { .. }
+            )
             && matches!(param.source, hir::GenericParamSource::Generics)
             && let parent_hir_id = self.tcx.parent_hir_id(hir_id)
             && let hir::Node::Item(item) = self.tcx.hir_node(parent_hir_id)
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_ssa/src/mir/block.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/place.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/coverageinfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/analyze.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/constant.rs" "/checkout/compiler/rustc_codegen_ssa/src/target_features.rs" "/checkout/compiler/rustc_codegen_ssa/src/mono_item.rs" "/checkout/compiler/rustc_codegen_ssa/src/codegen_attrs.rs" "/checkout/tests/run-make/libs-through-symlinks/bar.rs" "/checkout/tests/run-make/libs-through-symlinks/foo.rs" "/checkout/tests/run-make/exit-code/lint-failure.rs" "/checkout/tests/run-make/exit-code/success.rs" "/checkout/tests/run-make/exit-code/compile-error.rs" "/checkout/tests/run-make/exit-code/rmake.rs" "/checkout/tests/run-make/rust-lld-by-default-nightly/main.rs" "/checkout/tests/run-make/rust-lld-by-default-nightly/rmake.rs" "/checkout/tests/run-make/doctests-runtool/runtool.rs" "/checkout/tests/run-make/doctests-runtool/t.rs" "/checkout/tests/run-make/doctests-runtool/rmake.rs" "/checkout/tests/run-make/doctests-keep-binaries/t.rs" "/checkout/tests/run-make/doctests-keep-binaries/rmake.rs" "/checkout/compiler/rustc_passes/src/errors.rs" "/checkout/compiler/rustc_passes/src/diagnostic_items.rs" "/checkout/compiler/rustc_passes/src/stability.rs" "/checkout/compiler/rustc_passes/src/reachable.rs" "/checkout/compiler/rustc_passes/src/check_attr.rs" "/checkout/compiler/rustc_passes/src/loops.rs" "/checkout/compiler/rustc_passes/src/dead.rs" "/checkout/compiler/rustc_passes/src/lib.rs" "/checkout/compiler/rustc_passes/src/abi_test.rs" "/checkout/tests/run-make/macos-deployment-target/with_deployment_target.rs" "/checkout/compiler/rustc_passes/src/hir_id_validator.rs" "/checkout/compiler/rustc_passes/src/liveness/rwu_table.rs" "/checkout/compiler/rustc_passes/src/upvars.rs" "/checkout/compiler/rustc_passes/src/liveness.rs" "/checkout/compiler/rustc_passes/src/debugger_visualizer.rs" "/checkout/compiler/rustc_passes/src/check_const.rs" "/checkout/compiler/rustc_passes/src/lib_features.rs" "/checkout/compiler/rustc_passes/src/naked_functions.rs" "/checkout/compiler/rustc_passes/src/weak_lang_items.rs" "/checkout/compiler/rustc_passes/src/layout_test.rs" "/checkout/compiler/rustc_passes/src/lang_items.rs" "/checkout/compiler/rustc_passes/src/entry.rs" "/checkout/compiler/rustc_passes/src/hir_stats.rs" "/checkout/tests/run-make/lto-readonly-lib/main.rs" "/checkout/tests/run-make/lto-readonly-lib/lib.rs" "/checkout/tests/run-make/lto-readonly-lib/rmake.rs" "/checkout/tests/run-make/symbols-include-type-name/lib.rs" "/checkout/tests/run-make/symbols-include-type-name/rmake.rs" "/checkout/tests/run-make/ls-metadata/foo.rs" "/checkout/tests/run-make/ls-metadata/rmake.rs" "/checkout/tests/run-make/proc-macro-three-crates/c.rs" "/checkout/tests/run-make/proc-macro-three-crates/a.rs" "/checkout/tests/run-make/proc-macro-three-crates/rmake.rs" "/checkout/tests/run-make/proc-macro-three-crates/b.rs" "/checkout/tests/run-make/lto-dylib-dep/main.rs" "/checkout/tests/run-make/lto-dylib-dep/a_dylib.rs" "/checkout/tests/run-make/lto-dylib-dep/rmake.rs" "/checkout/compiler/rustc_errors/src/registry.rs" "/checkout/tests/run-make/alloc-no-rc/rmake.rs" "/checkout/compiler/rustc_errors/src/json/tests.rs" "/checkout/compiler/rustc_errors/src/emitter.rs" "/checkout/compiler/rustc_errors/src/codes.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/statement.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sun Aug 18 13:29:22 UTC 2024
  network time: Sun, 18 Aug 2024 13:29:23 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10441070872/job/28911759874) [(plain)](https://github.com/rust-lang/rust/commit/50691287d597330457a8135ca6b6d9601f9a1187/checks/28911759874/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/lint/must_not_suspend/other_items.rs stdout ----
diff of stderr:

- error: `must_not_suspend` attribute should be applied to a struct, enum, or trait
+ error: `must_not_suspend` attribute should be applied to a struct, enum, union, or trait
2   --> $DIR/other_items.rs:5:1
4 LL | #[must_not_suspend]

5    | ^^^^^^^^^^^^^^^^^^^
6 LL | mod inner {}
---
To only update this specific test, also pass `--test-args lint/must_not_suspend/other_items.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/must_not_suspend/other_items.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/must_not_suspend/other_items" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/must_not_suspend/other_items/auxiliary" "--edition=2018"
--- stderr -------------------------------
--- stderr -------------------------------
error: `must_not_suspend` attribute should be applied to a struct, enum, union, or trait
   |
   |
LL | #[must_not_suspend] //~ ERROR attribute should be
LL | mod inner {}
   | ------------ is not a struct, enum, union, or trait

error: aborting due to 1 previous error
error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/lint/must_not_suspend/return.rs stdout ----
diff of stderr:

- error: `must_not_suspend` attribute should be applied to a struct, enum, or trait
+ error: `must_not_suspend` attribute should be applied to a struct, enum, union, or trait
3    |
4 LL |   #[must_not_suspend]

6 LL | / fn foo() -> i32 {
---
To only update this specific test, also pass `--test-args lint/must_not_suspend/return.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/must_not_suspend/return.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/must_not_suspend/return" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/must_not_suspend/return/auxiliary" "--edition=2018"
--- stderr -------------------------------
--- stderr -------------------------------
error: `must_not_suspend` attribute should be applied to a struct, enum, union, or trait
   |
   |
LL |   #[must_not_suspend] //~ ERROR attribute should be
LL | / fn foo() -> i32 {
LL | |     0
LL | | }
   | |_- is not a struct, enum, union, or trait
```

</details>

GoldsteinE committed:
Check that `#[may_dangle]` is properly applied

It's only valid when applied to a type or lifetime parameter
in `Drop` trait implementation.

GoldsteinE committed:
Fix wording of misapplied `must_not_suspend` error

compiler-errors commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -1373,6 +1373,28 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
 
+    /// Checks if `#[may_dangle]` is applied to a lifetime or type generic parameter in `Drop` impl.
+    fn check_may_dangle(&self, hir_id: HirId, attr: &Attribute, target: Target) {
+        if matches!(target, Target::GenericParam(_))
```
this first `matches` is redundant if we are just checking the node one line below (1379)

GoldsteinE replied:
Fair, I removed this check.

compiler-errors commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -1373,6 +1373,28 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
+            )
+            && matches!(param.source, hir::GenericParamSource::Generics)
+            && let parent_hir_id = self.tcx.parent_hir_id(hir_id)
+            && let hir::Node::Item(item) = self.tcx.hir_node(parent_hir_id)
+            && let hir::ItemKind::Impl(impl_) = item.kind
+            && let Some(trait_) = impl_.of_trait
```
```suggestion
            && let hir::Node::Item(hir::Item { kind: hir::ItemKind::Impl(impl_), .. }) = self.tcx.hir_node(parent_hir_id)
            && let Some(trait_) = impl_.of_trait
```

GoldsteinE replied:
This leads to somewhat ugly formatting:

```rust
&& let hir::Node::Item(hir::Item { kind: hir::ItemKind::Impl(impl_), .. }) =
    self.tcx.hir_node(parent_hir_id)
```

I think the current version is nicer, but I don’t feel that strong about it, so I can make this change if you prefer it.

compiler-errors reviewed (approved):
Couple tiny tweaks, otherwise LGTM

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit df6cb954bb50fa8567f7198c02db90edda3cd3e4 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors df6cb954bb50fa8567f7198c02db90edda3cd3e4 -->
<!-- homu: {"type":"Approved","sha":"df6cb954bb50fa8567f7198c02db90edda3cd3e4","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
